payjoin/src/core/receive/mod.rs

```diff
@@ -747,27 +767,114 @@ pub(crate) mod tests {
+            p2tr_pair.err().unwrap(),
+            PsbtInputError::from(InternalPsbtInputError::from(InputWeightError::NotSupported))
         );
 
+        // If UTXO is non-P2TR
         let p2sh_txout = TxOut {
             value: Default::default(),
             script_pubkey: ScriptBuf::new_p2sh(&ScriptHash::all_zeros()),
         };
         assert_eq!(
-            InputPair::new_p2tr(p2sh_txout, outpoint, Some(sequence)).err().unwrap(),
+            InputPair::new_p2tr_keyspend(p2sh_txout, outpoint, Some(sequence)).err().unwrap(),
             PsbtInputError::from(InvalidScriptPubKey(AddressType::P2tr))
         )
     }
 
+    #[test]
+    fn p2tr_expected_weight_from_witness() {
+        let outpoint = OutPoint { txid: Txid::from_byte_array(DUMMY32), vout: 31 };
+        let sequence = Sequence::ZERO;
+        let pubkey_string = "0347ff3dacd07a1f43805ec6808e801505a6e18245178609972a68afbc2777ff2b";
+        let pubkey = pubkey_string.parse::<PublicKey>().expect("valid pubkey");
+        let xonly_pubkey = XOnlyPublicKey::from(pubkey.inner);
+        let p2tr_txout = TxOut {
+            value: Amount::from_sat(12345),
+            script_pubkey: ScriptBuf::new_p2tr(SECP256K1, xonly_pubkey, None),
+        };
+        let base_txin = TxIn { previous_output: outpoint, sequence, ..Default::default() };
+        let psbtin = psbt::Input { witness_utxo: Some(p2tr_txout.clone()), ..Default::default() };
+
+        // Key-path spend with default sighash (64-byte signature)
+        let mut key_witness = witness::Witness::new();
+        key_witness.push(vec![0x01; 64]);
+        let txin = TxIn { witness: key_witness.clone(), ..base_txin.clone() };
+        let key_input_weight = Weight::from_non_witness_data_size(txin.base_size() as u64)
+            + Weight::from_witness_data_size(key_witness.size() as u64);
+        let pair =
+            InputPair::new(txin, psbtin.clone(), None).expect("taproot key witness provided");
+        assert_eq!(pair.expected_weight, key_input_weight);
+
+        // Key-path spend with non-default sighash (65-byte signature)
+        let mut sighash_witness = witness::Witness::new();
+        let mut signature = vec![0x02; 65];
+        signature[64] = 0x01;
+        sighash_witness.push(signature);
+        let txin = TxIn { witness: sighash_witness.clone(), ..base_txin.clone() };
+        let input_weight = Weight::from_non_witness_data_size(txin.base_size() as u64)
+            + Weight::from_witness_data_size(sighash_witness.size() as u64);
+        let pair = InputPair::new(txin, psbtin.clone(), None)
+            .expect("taproot non-default sighash witness");
+        assert_eq!(pair.expected_weight, input_weight);
+
+        // Script-path spend (multiple witness elements)
+        let mut script_witness = witness::Witness::new();
+        script_witness.push(vec![0x03; 64]);
+        script_witness.push(vec![0x04; 5]);
+        script_witness.push(vec![0x05; 33]);
```

Comment:
I think parts of this test are redundant. `expected_input_weight` treats all witness elements the same. i.e the default and non-default sig test case will hit the same code path as the multi-witness element test.